README.md
# Ccsh

[![Build Status](https://travis-ci.org/raffs/ccsh.svg?branch=master)](https://travis-ci.org/raffs/ccsh)
[![Gem Version](https://badge.fury.io/rb/ccsh.svg)](https://badge.fury.io/rb/ccsh)


ccsh is an interactive shell console connected into multiple servers, the intent
of this interface is to simplify the need to run very simple task on the
machine instances.

ccsh will read the servers from the **host** file and create an interactive shell
for running commands on all selected servers

Project Information:
* **Rubygem**: [gems/ccsh](https://rubygems.org/gems/ccsh)
* **Documentation** [raffs.github.io/ccsh](https://raffs.github.io/ccsh)

## Setup

### Requirements

* Ruby
* Rubygem

#### Installation

Install the ccsh gem
```sh
gem install ccsh
```

#### Create host file

By default the ccsh command will load the file from ```~/.cssh/host.yaml```. The file path could change using the
```ccsh -h path/to/HOSTFILE``` command.

```yaml
---
# Example: hosts_exa01.yaml
#
# Default example from the ccsh connections
version: 1.0

# define default settings for each defined host
defaults:
  port: 22
  user: user
  password: password
  private_key: ~/.ssh/id_rsa
  ssh_options:
    timeout: 720
    host_key: 'ssh-rsa'

# list of host file
hosts:

  - name: Node 01
    hostname: node01.example.com
    groups:
      - production
      - webserver

  - name: Node 02
    hostname: node02.example.com
    groups:
      - production
      - dbserver
```

### Start the console

Now, the hosts file are set, we can start the ccsh console given the group name
such as ```all``` or ```webserver```.

```bash
$ ccsh webserver
```

connected to the ccsh console we can type the commands we want to perform on
each connected host.

```bash
ccsh> uptime
>>> node01.example.com {"rc"=>0, "time"=>2.941396}
 03:13:20 up 85 days, 15:14,  1 user,  load average: 0.21, 0.07, 0.03

>>> node02.example.com {"rc"=>0, "time"=>2.941975}
 03:13:20 up 85 days, 11:14,  1 user,  load average: 0.33, 0.10, 0.07
```

## Contributing

* [Github Issues](https://github.com/raffs/ccsh/issues)
* [Pull Requests](https://github.com/raffs/ccsh/pulls)

### Testing

To run the rspec spec run:
```sh
rake spec
```

### Build

To build the ccsh gem package run:
```sh
gem build ccsh.gemspec
```

### Installing local gem

To install the generated gem package run:
```sh
gem install_local gem-VERSION.gem
```

Optionally you can run the *build* and *install locally* with the `rake deploy_local`

### Process

* **Fork** the repository
* Create a **branch** following the pattern: feature/<FEATURE_SHORT_DESCRIPTION>
	* Implement the new feature
	* Add spec test for each implemented task/feature/function
* Create the **Pull Request**


CHANGELOG.md
# CCSH Changelog

All notables changes to the CCSH project will be documented in this file

## [1.0.0] - Unreleased

### [0.0.5] - Unreleased

#### Added

- (Issue: [#18]) - Investigate why the sudo command it's taking to long to execute.
- (Issue: [#21]) - Added validation code to empty targets and filters.
- (Issue: [#19]) - Improve the CI Process (tag, PR's, branches and travis deployments).
- (Issue: [#16]) - Add a enable_sudo, sudo_user and sudo_password options.

#### Fixes

- (Issue: [#17]) - Add a validation code to quit the system when there's no group or servers matched.

### [0.0.4] - Jan 19, 2018

#### Added

- (Issue: [#8])  - Handle ruby errors with better and understandable messages
- (Issue: [#4])  - Implement the max_threads options
- (Issue: [#12]) - Add --show-hosts option.
- (Issue: [#10]) - Handle more then one group or host filter.
- (Issue: [#7])  - Added the output to file option.

#### Fixes

- (Issue: [#13]) - Fix the verbose command not showing the target messages.
- (Issue: [#9]) - Fix special characters on the gem authors name.

### [0.0.3] - Dec 29, 2017

#### Fixes

- (Issue: [#6]) - Fix the default option that points to user's home directory.

### [0.0.2] - Dec 27, 2017

#### Added

- (Issue: [#3]) - Use ssh_options for parsing ruby SSH extended options.
- (Issue: [#2]) - Enable the Usage of private_key instead of only password.
- (Issue: [#1]) - Use threads to execute the ssh connections.

# That's all folks

[0.0.2]: https://github.com/raffs/ccsh/compare/v0.0.1...v0.0.2
[0.0.3]: https://github.com/raffs/ccsh/compare/v0.0.2...v0.0.3
[0.0.4]: https://github.com/raffs/ccsh/compare/v0.0.4...v0.0.3
[0.0.5]: https://github.com/raffs/ccsh/compare/v0.0.5...v0.0.4
[#1]: https://github.com/raffs/ccsh/issues/1
[#2]: https://github.com/raffs/ccsh/issues/2
[#3]: https://github.com/raffs/ccsh/issues/3
[#4]: https://github.com/raffs/ccsh/issues/4
[#6]: https://github.com/raffs/ccsh/issues/6
[#7]: https://github.com/raffs/ccsh/issues/7
[#8]: https://github.com/raffs/ccsh/issues/8
[#9]: https://github.com/raffs/ccsh/issues/9
[#10]: https://github.com/raffs/ccsh/issues/10
[#12]: https://github.com/raffs/ccsh/issues/12
[#13]: https://github.com/raffs/ccsh/issues/13
[#16]: https://github.com/raffs/ccsh/issues/16
[#17]: https://github.com/raffs/ccsh/issues/17
[#18]: https://github.com/raffs/ccsh/issues/18
[#19]: https://github.com/raffs/ccsh/issues/19
[#21]: https://github.com/raffs/ccsh/issues/21


spec/ccsh/hosts.rb
# require 'spec_helper'
#
# describe CCSH::Hosts do
#     let { :default_file } {
#
#     }
#
#     context "#parser!" do
#
#     end
#
#     context "#filter_by" do
#
#     end
# end


spec/spec_helper.rb
require 'ccsh'

RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/ccsh/utils_spec.rb
require 'spec_helper'

describe CCSH::Utils do

    context '#get_options' do
        let ( :host ) {{
            'name' => 'example',
        }}

        it 'should return the host value' do
            expect(CCSH::Utils.get_options('name', host, 'default_example')).to eq 'example'
        end

        it 'should return the default value' do
            expect(CCSH::Utils.get_options('hostname', host, 'default_hostname')).to eq 'default_hostname'
        end
    end

    context "#verbose" do
        let ( :message ) { 'ccsh message' }

        describe "turn on verbose mode" do
            it 'should display message' do
                allow(ENV).to receive(:[]).with('CCSH_VERBOSE').and_return('true')
                expect{ CCSH::Utils.verbose message }.to output(/ccsh message/).to_stdout
            end
        end

        describe "turn off verbose mode" do
            it 'should not display message' do
                allow(ENV).to receive(:[]).with('CCSH_VERBOSE').and_return('false')

                expect{ CCSH::Utils.verbose message }.to output(//).to_stdout
                expect{ CCSH::Utils.verbose message }.to_not output(/ccsh message/).to_stdout
            end
        end
    end

    context "#debug" do
        let ( :message ) { 'debug message' }

        describe "turn on verbose mode" do
            it 'should display message' do
                allow(ENV).to receive(:[]).with('CCSH_DEBUG').and_return('true')

                expect{ CCSH::Utils.debug message }.to output(/debug message/).to_stdout
            end
        end

        describe "turn off verbose mode" do
            it 'should not display message' do
                allow(ENV).to receive(:[]).with('CCSH_DEBUG').and_return('false')

                expect{ CCSH::Utils.debug message }.to output(//).to_stdout
                expect{ CCSH::Utils.debug message }.to_not output(/debug message/).to_stdout
            end
        end
    end
end

lib/ccsh.rb
require 'rubygems' unless defined?(Gem)

module CCSH
    require 'ccsh/version.rb'
    require 'ccsh/utils.rb'
    require 'ccsh/options.rb'
    require 'ccsh/host'
    require 'ccsh/hosts.rb'
    require 'ccsh/ssh.rb'

    def self.execute!
        begin
            options = CCSH::Options.parse_options ARGV

            ENV['CCSH_DEBUG']   = "true" if options[:debug]
            ENV['CCSH_VERBOSE'] = "true" if options[:verbose]

            if options[:targets].empty?
                error_msg = "You must specify a target hostname or group name."

                puts error_msg
                puts "Example:"
                puts "    'ccsh databases' - select all servers from the database group"
                puts "    'ccsh all'       - select all defined servers"
                puts "Please run 'ccsh --help' for details"
                puts ""

                raise error_msg
            end

            filename = options[:hosts]
            targets = options[:targets]
            hosts = CCSH::Hosts.new.parser!(filename).filter_by(targets)

            # validate and display pretty message if no hosts was found to the
            # given target.
            if hosts.empty?
                error_msg = "Could not found any hosts that matches the given target(s): '#{targets.join(', ')}'"
                puts error_msg

                puts "Here are some groups found on the file: '#{filename}'"
                CCSH::Utils.show_groups(CCSH::Hosts.new.parser!(filename).filter_by('all'))

                puts ""
                raise error_msg
            end

            self.start_cli(hosts, options)

        rescue Exception => e
            if not e.message == 'exit'
                CCSH::Utils.debug "Backtrace:\n\t#{e.backtrace.join("\n\t")}\n\n"
                CCSH::Utils.verbose "Backtrace:\n\t#{e.backtrace.join("\n\t")}\n\n"

                STDERR.puts "An error occur and system exit with the following message: "
                STDERR.puts "  #{e.message.gsub("\n", ' ').squeeze(' ')}"

                exit!
            end

            exit
        end
    end

    def self.with_info(options)
        cmd_start = Time.now
        cmd = yield
        cmd_end = Time.now

        info = {
            'rc'   => cmd.return_code,
            'time' => cmd_end - cmd_start,
        }.inspect

        prompt = ">>> #{cmd.hostname} #{info}"

        puts prompt
        STDERR.puts cmd.stderr if cmd.stderr != nil && cmd.stderr != ''
        puts cmd.stdout if cmd.stdout != nil && cmd.stdout != ''
        puts

        if options[:output] != nil
            begin
                file_handler = File.new(options[:output], "a+")

                file_handler.write("#{prompt}\n")
                file_handler.write("ERROR: #{cmd.stderr}") if cmd.stderr != nil && cmd.stderr != ''
                file_handler.write("#{cmd.stdout}\n") if cmd.stdout != nil && cmd.stdout != ''
            rescue Exception => e
                puts "WARNING: An error occur when trying to write the output to file: #{options[:output]}. "
                puts "WARNING: Thefore, the output is not being stored"
                puts "WARNING: Error message: #{e.message}"
            end
        end
    end


    def self.start_cli(hosts, options)
        CCSH::Utils.display_hosts(hosts) if options[:show_hosts]
        CCSH::Utils.display_hosts_verbosity(hosts) if options[:verbose]

        quit = false
        loop do
            begin
                CCSH::Utils.handle_signals
                printf "ccsh> "
                command = STDIN.gets.chomp

                if command != ''
                    CCSH::Utils.exit_console 0 if command == 'quit'
                    CCSH::Utils.exit_console 0 if command == 'exit'

                    if command == 'clear'
                        CCSH::Utils.clear_console
                    elsif command == 'reset'
                        CCSH::Utils.reset_console
                    else
                        if ((options.max_threads == 0) || (options.max_threads > hosts.length))
                            options.max_threads = hosts.length
                        end

                        CCSH::Utils.debug "Using #{options.max_threads} maximum of threads"

                        hosts.each_slice(options.max_threads) do |batch_hosts|
                            threads = []

                            batch_hosts.each do |host|
                                if command =~ /^([ ]*)sudo/
                                    if host.sudo_enabled != true
                                        STDERR.puts "WARN: You cannot run sudo on the host #{host.hostname}, please enable sudo mode for this hosts"
                                        next
                                    end
                                end

                                threads << Thread.new do
                                    info = with_info(options) do
                                        host.run command
                                    end
                                end
                            end

                            threads.each(&:join)
                        end
                    end
                end
            rescue Exception => exception
                STDERR.puts exception.message
                STDERR.puts
                raise exception
            end
        end
    end
end


lib/ccsh/ssh.rb
module CCSH
    module SSH
        require 'net/ssh'

        def self.start
            cmd = RemoteCommand.new
            yield cmd

            raise 'Could not parser the remote command' if cmd.command == nil
            raise 'Could not parser the hostname' if cmd.hostname == nil

            return cmd.execute!
        end

        class RemoteCommand
            attr_accessor :command
            attr_accessor :hostname
            attr_accessor :options
            attr_accessor :user
            attr_accessor :port
            attr_accessor :password
            attr_accessor :private_key
            attr_accessor :options
            attr_accessor :enable_sudo
            attr_accessor :sudo_password

            attr_accessor :stdout
            attr_accessor :stderr
            attr_accessor :return_code
            attr_accessor :return_signal

            def initialize
                @command  = nil
                @hostname = nil
                @options  = []
                @stdout = ''
                @stderr = ''

                @enable_sudo      = false
                @sudo_user        = 'root'
                @sudo_password    = nil
            end

            def execute!
                @options = {
                    :password => @password,
                    :keys     => [@private_key],
                    :port     => @port,
                    :host_key => @options['host_key'],
                    :timeout  => @options['timeout'],
                }.select {|key,value| value != nil}

                raise "something" unless CCSH::Utils.valid_ssh(@options)

                Net::SSH.start(@hostname, @user, @options) do |ssh|
                    ssh.open_channel do |ch|
                        ch.on_data do |c, data|

                            if data =~ /^\[sudo\] password for /

                                if @enable_sudo
                                    if @sudo_password != nil
                                        ch.send_data "#{@sudo_password}\n"
                                    else
                                        msg = """The server #{@hostname} asked for user password
                                            by the 'password' or 'sudo_password' option was not defined.
                                        """.gsub("\n", ' ').squeeze(' ')

                                        STDERR.puts "[WARN] #{msg}"
                                        ch.send_data "\x03"
                                    end
                                else
                                    msg = """The server #{@hostname} asked for user password
                                        the sudo_enabled option was not allowed
                                        skipping this hosts setup
                                    """.gsub("\n", ' ').squeeze(' ')

                                    STDERR.puts "[WARN] #{msg}"
                                    ch.send_data "\x03"
                                end
                            else
                                @stdout << data
                            end
                        end

                        ch.on_extended_data do |c, type, data|
                            @stderr << data
                        end

                        ch.on_request("exit-status") do |c,data|
                            @return_code = data.read_long
                        end

                        ch.on_request("exit-signal") do |c, data|
                            @return_signal = data.read_long
                        end

                        ch.request_pty do |ch, success|
                            if success
                                CCSH::Utils.debug("pty successfully obtained for #{@hostname}")
                            else
                                CCSH::Utils.debug("pty unsuccessfully obtained for #{@hostname}")
                            end
                        end

                        ch.exec @command do |ch, success|
                            raise "Could execute command #{command} on #{host}" unless success
                        end
                    end.wait
                end

                return self
            end
        end
    end
end


lib/ccsh/host.rb
module CCSH
    class Host
        attr_accessor :name
        attr_accessor :port
        attr_accessor :user
        attr_accessor :groups
        attr_accessor :hostname
        attr_accessor :password
        attr_accessor :private_key
        attr_accessor :ssh_options
        attr_accessor :timeout
        attr_accessor :sudo_enabled
        attr_accessor :sudo_password

        def initialize
            @name = ''
            @port = 22
            @user = 'root'
            @groups = ''
            @hostname = '127.0.0.1'
            @password = nil
            @private_key = nil
            @ssh_options = []
            @sudo_enabled = false
            @sudo_password = @password
        end

        def run(command)
            CCSH::Utils.debug("Running command '#{command}' on #{@hostname}")

            command = CCSH::SSH.start do |ssh|
                ssh.user = @user
                ssh.port = @port
                ssh.options = @ssh_options
                ssh.password = @password
                ssh.private_key = @private_key
                ssh.options = @ssh_options
                ssh.enable_sudo = @sudo_enabled
                ssh.sudo_password = @sudo_password

                ssh.command = command
                ssh.hostname = @hostname || @name
            end

            return command
        end

        def contain?(target)
            return @groups.include? target
        end
    end
end


lib/ccsh/hosts.rb
module CCSH
    require 'yaml'
    require 'pathname'

    require 'ccsh/host'

    class Hosts
        attr_accessor :defaults
        attr_accessor :hosts

        def initialize
            # TODO
        end

        def parser!(filename)
            begin
                file = Pathname.new(filename)

                raise "The file does not exists: '#{file.realpath}'" if not file.exist?
                raise "The host file is not regular file: '#{file.realpath}'" if not file.file?

                hostfile = YAML.load_file(file.realpath)

                @hosts = []
                @defaults = CCSH::Utils.merge_defaults(hostfile['defaults'])

                hostfile['hosts'].each do |h|
                    host = Host.new

                    host.name = CCSH::Utils.get_options('name', h, @defaults['name'])
                    host.user = CCSH::Utils.get_options('user', h, @defaults['user'])
                    host.port = CCSH::Utils.get_options('port', h, @defaults['port'])

                    host.hostname = CCSH::Utils.get_options('hostname', h, @defaults['hostname'])
                    host.password = CCSH::Utils.get_options('password', h, @defaults['password'])
                    host.private_key = CCSH::Utils.get_options('private_key', h, @defaults['private_key'])
                    host.ssh_options = CCSH::Utils.get_options('ssh_options', h, @defaults['ssh_options'])
                    host.timeout = CCSH::Utils.get_options('timeout', h, @defaults['timeout'])
                    host.sudo_enabled = CCSH::Utils.get_options('sudo_enabled', h, @defaults['sudo_enabled'])
                    host.sudo_password = CCSH::Utils.get_options('sudo_password', h, @defaults['sudo_password'])

                    # define the password if the sudo_password is not set
                    host.sudo_password = host.password if host.sudo_password == nil

                    host.groups = h['groups']

                    @hosts << host
                end
            rescue Exception => e
                puts e.message
                puts e.backtrace
                puts
            end

            return self
        end

        def filter_by(targets)
            return @hosts if targets.include?('all')

            hosts = []
            @hosts.each do |host|
                targets.each do |target|
                    hosts << host if host.groups != nil && host.groups.include?(target)
                end
            end

            return hosts
        end
    end
end


lib/ccsh/options.rb
module CCSH
    module Options
        require 'ostruct'
        require 'optparse'

        def self.parse_options(args)
            options = OpenStruct.new

            # default values
            user_home = File.expand_path('~')

            options.config      = "#{user_home}/.ccsh/config.yaml"
            options.hosts       = "#{user_home}/.ccsh/hosts.yaml"
            options.output      = ""
            options.debug       = false
            options.verbose     = false
            options.check       = false
            options.output      = nil
            options.show_hosts  = false
            options.max_threads = 0

            # open parser
            opt_parser = OptionParser.new do |opts|
                opts.banner = "Usage: ccsh [options] GROUP1 GROUP2 ... "
                opts.separator ""
                opts.separator "Options: "

                opts.on("-c", "--config CONFIG_FILE", "Configuration file (default: ~/.ccsh/config.yaml)") do |cfg|
                    options.config = cfg
                end

                opts.on("-d", "--[no-]debug", "Run Debug mode") do |d|
                    options.verbose = d
                    options.debug = d
                end

                opts.on("-h", "--hosts HOST_FILE", "Specified hosts file") do |h|
                    options.hosts = h
                end

                opts.on("-k", "--[no-]check", "Check host connection before continuing") do |k|
                    options.check = k
                end

                opts.on("-o", "--output FILEPATH", "Write all the interaction into an output file") do |out|
                    options.output = out
                end

                opts.on("-v", "--[no-]verbose", "Run verbosely") do |v|
                    options.show_hosts = v
                    options.verbose = v
                end

                opts.on("--show-hosts", "Output the filtered hosts") do |show_hosts|
                    options.show_hosts = show_hosts
                end

                opts.on("-t", "--max-threads MAX_THREADS",
                        "Define the maxinum number of threads (by default: the ammount of selected hosts)") do |max_threads|

                    begin
                        if not max_threads =~ /\d+/
                            raise "The value '#{max_threads}' is not a valid value to max_threads options"
                        end

                        max_threads = max_threads.to_i
                        if max_threads < 0
                            raise "max_threads value should be a positive number instead of '#{max_threads}'"
                        end

                        options.max_threads = max_threads
                    rescue Exception => e
                        raise e.message
                    end
                end

                opts.on("--version", "Display version") do |v|
                    puts "CCSH version #{CCSH::VERSION}"
                    exit
                end
            end
            opt_parser.parse!(args)

            options.targets = args
            return options
        end
    end
end


lib/ccsh/utils.rb
module CCSH
    module Utils

        ##
        # Responsible to merge the defaults settings with the each hosts
        # parameters.
        def self.merge_defaults(defaults)
            defaultsValues = {
                'user'          => 'root',
                'port'          => '22',
                'sudo_enabled'  => false,
                'sudo_password' => nil,
                'private_key'   => '~/.ssh/id_rsa',
            }
            defaultsValues.merge!(defaults) if defaults != nil

            ssh_options = {
                'timeout' => 720,
                'ssh-rsa' => 'ssh-rsa',
            }

            if defaults['ssh_options'] != nil
                defaultsValues['ssh_options'] = ssh_options.merge!(defaults['ssh_options'])
            else
                defaultsValues['ssh_options'] = ssh_options
            end

            return defaultsValues
        end

        ##
        # Given an item, hash and wanted default return, return the
        # item if exists on array object or return the default
        # value if the key doesn't exist.
        def self.get_options(item, object, default)
            if object.key? item
                return object[item]
            end

            return default
        end

        ##
        # Pretty print the filtered hosts
        def self.display_hosts(hosts)
            puts "Selected Hosts (option: --show-hosts):"
            hosts.each do |host|
                puts "+ Server #{host.name}: #{host.user}@#{host.hostname} => Groups: #{host.groups}"
            end
            puts
        end

        ##
        # Display host display_hosts_verbosity
        def self.display_hosts_verbosity(hosts)
            hosts.each do |host|
                CCSH::Utils::verbose("sudo mode enable to host: #{host.hostname}") if host.sudo_enabled
            end
            puts
        end

        ##
        # Show at least 10 groups names
        def self.show_groups(hosts)
            hosts_groups = []
            hosts.each do |host|
                if host.groups != nil
                    host.groups.each do |gname|
                        hosts_groups.push(gname) unless hosts_groups.include?(gname)
                    end
                end
            end

            hosts_groups.slice(0, 10).each { |gname| puts "  - #{gname}" }
            puts "Example: ccsh #{hosts_groups[0]}" unless hosts_groups[0] == nil
            puts "Example: ccsh #{hosts_groups.slice(0,2).join(' ')}" unless hosts_groups.slice(0,2) == nil
        end

        ##
        # validate the ssh connection before start the ccsh console
        def self.valid_ssh(options)
            return true
        end

        ##
        # Display a message, if debug mode is enabled
        def self.debug(msg)
            puts "[DEBUG] #{msg}" if ENV['CCSH_DEBUG'] == "true"
        end

        ##
        # Verbosely display a message, if verbose mode is enabled
        def self.verbose(msg)
            puts "[verbose] #{msg}" if ENV['CCSH_VERBOSE'] == "true"
        end

        ##
        # Handle terminal signals
        def self.handle_signals
            Signal.trap('INT') do
                self.exit_console 0
            end

            Signal.trap('TERM') do
                self.exit_console 0
            end
        end

        ##
        # Exit the console message
        def self.exit_console(code, msg = nil)
            puts msg if msg != nil

            puts "\nBye..."
            exit code || 0
        end

        ##
        # clear the system console when run clear
        def self.clear_console
            printf "\e[H\e[2J"
        end

        ##
        # reset the console
        def self.reset_console
            printf("\033c");
        end
    end
end
